src/transformers/tokenization_utils.py

```diff
@@ -49,6 +49,126 @@
+        The special key `""` is used to represent termination.
+
+        This function is idempotent, adding twice the same word will leave the trie unchanged
+
+        >>> trie = Trie()
+
+        >>> trie.add("Hello 友達")
+
+        >>> trie.data
+
+        {"H": {"e": {"l": {"l": {"o": {" ": {"友": {"達": {"": 1}}}}}}}}}
+
+        >>> trie.add("Hello")
+
+        >>> trie.data
+
+        {"H": {"e": {"l": {"l": {"o": {"": 1, " ": {"友": {"達": {"": 1}}}}}}}}}
+        """
+        if not word:
+            # Prevent empty string
+            return
+        ref = self.data
+        for char in word:
+            ref[char] = char in ref and ref[char] or {}
+            ref = ref[char]
+        ref[""] = 1
+
+    def split(self, text: str) -> List[str]:
+        """
+        Will look for the words added to the trie within `text`. Output is the original string splitted along the
+        boundaries of the words found.
+
+        This trie will match the longest possible word first !
+
+        >>> trie = Trie()
+
+        >>> trie.split("[CLS] This is a extra_id_100")
+
+        ["[CLS] This is a extra_id_100"]
+
+        >>> trie.add("[CLS]")
+
+        >>> trie.add("extra_id_1")
+
+        >>> trie.add("extra_id_100")
+
+        >>> trie.split("[CLS] This is a extra_id_100")
+
+        ["[CLS]", " This is a ", "extra_id_100"]
+        """
+        states = {}
+        offsets = []
+
+        skip = None
+        for current, c in enumerate(text):
+            if skip and current < skip:
+                # Prevents the lookahead for matching twice
+                # like extra_id_100 and id_100
+                continue
+            to_remove = set()
+            reset = False
+            for start, p in states.items():
+                if c in p:
+                    p = p[c]
+                    if "" in p:
+                        # Lookahead to match longest first
+                        # Important in case of extra_id_1 vs extra_id_100
+                        lookahead_index = current + 1
+                        end = current + 1
+                        next_c = c
+                        while next_c in p:
+                            p = p[next_c]
+                            lookahead_index += 1
+                            if "" in p:
+                                end = lookahead_index + 1
+                                skip = lookahead_index + 1
+
+                            if lookahead_index == len(text):
+                                # End of string
+                                break
+                            next_c = text[lookahead_index]
+                        # End lookahead
+
+                        offsets.append(start)
+                        offsets.append(end)
+                        reset = True
+                    states[start] = p
+                else:
+                    to_remove.add(start)
```

Comment:
Is there any way to replace the variable name by something more explicit? This is super hard to read and understand.
More comments would also be welcome.

Reply:
I can change `c` into `current_char` and `p` into `trie_pointer` maybe ?

Overall state machines code is always a bit hard to read (and complexified by the lookahead part).
I don't know a great way to make this a usual python-readable nice.

I think the best way to improve readability is to document a bit more what each variable should be, or do you think variable names are the bigger culprits ?

Reply:
At least knowing c is a char helps tremendously, so all for the variable renames! And a few comments on what is happening would be amazing as well if you can.

Reply:
I added lots of comments everywhere, hopefully explaining a bit more how every piece is supposed to work and interact with each other.

Reply:
Thanks, it's very helpful!